src/hx/libs/regexp/RegExp.cpp

```diff
@@ -34,76 +43,92 @@ struct pcredata : public hx::Object
       HX_OBJ_WB_GET(this, expr.raw_ref());
       flags = inFlags;
 
-
-      nmatchs = 0;
-      pcre_fullinfo(rUtf8,NULL,PCRE_INFO_CAPTURECOUNT,&nmatchs);
-      nmatchs++;
-      matchs = (int*)malloc(sizeof(int) * 3 * nmatchs);
+      n_groups = 0;
+      pcre2_pattern_info_8(rUtf8,PCRE2_INFO_CAPTURECOUNT,&n_groups);
+      n_groups++;
+      match_data8 = pcre2_match_data_create_from_pattern_8(rUtf8, NULL);
+      #ifdef HX_SMART_STRINGS
+      match_data16 = 0;
+      #endif
 
       _hx_set_finalizer(this, finalize);
    }
 
    #ifdef HX_SMART_STRINGS
-   void create16(pcre16 *inR, String inExpr, int inFlags)
+   void create16(pcre2_code_16 *inR, String inExpr, int inFlags)
    {
       rUtf8 = 0;
       rUtf16 = inR;
       expr = inExpr;
       HX_OBJ_WB_GET(this, expr.raw_ref());
       flags = inFlags;
 
-      nmatchs = 0;
-      pcre16_fullinfo(rUtf16,NULL,PCRE_INFO_CAPTURECOUNT,&nmatchs);
-      nmatchs++;
-      matchs = (int*)malloc(sizeof(int) * 3 * nmatchs);
+      n_groups = 0;
+      pcre2_pattern_info_16(rUtf16,PCRE2_INFO_CAPTURECOUNT,&n_groups);
+      n_groups++;
+      match_data8 = 0;
+      match_data16 = pcre2_match_data_create_from_pattern_16(rUtf16, NULL);
 
       _hx_set_finalizer(this, finalize);
    }
    #endif
 
-
    bool run(String string,int pos,int len)
    {
       #ifdef HX_SMART_STRINGS
       if (string.isUTF16Encoded())
       {
          if (!rUtf16)
          {
-            const char *error = 0;
-            int err_offset = 0;
+            int error_code;
+            size_t error_offset;
             hx::strbuf buf;
-            rUtf16 = pcre16_compile((PCRE_SPTR16)expr.wc_str(&buf),flags|PCRE_UTF16,&error,&err_offset,NULL);
-            if (!rUtf16)
-            {
-               return false;
+            rUtf16 = pcre2_compile_16((PCRE2_SPTR16)expr.wc_str(&buf),expr.length,flags,&error_code,&error_offset,NULL);
+            if (!rUtf16) {
+               regexp_compilation_error(expr,error_code,error_offset);
             }
+            match_data16 = pcre2_match_data_create_from_pattern_16(rUtf16, NULL);
          }
 
-         int n =  pcre16_exec(rUtf16,NULL,(const unsigned short *)string.raw_wptr(),pos+len,pos,PCRE_NO_UTF16_CHECK,matchs,nmatchs * 3);
+         int n = pcre2_match_16(rUtf16,(PCRE2_SPTR16)string.raw_wptr(),pos+len,pos,PCRE2_NO_UTF_CHECK,match_data16,NULL);
          return n>=0;
       }
 
       if (!rUtf8)
       {
-         rUtf8 =  pcre_compile(expr.utf8_str(),flags|PCRE_UTF8,0,0,0);
-         if (!rUtf8)
-            return false;
+         int error_code;
+         size_t error_offset;
+         int utf8Length = 0;
+         PCRE2_SPTR8 utf8 = (PCRE2_SPTR8)expr.utf8_str(NULL, true, &utf8Length);

```

Comment:
@hughsando When converting utf16 to utf8 here, it is done without a manually created buffer, unlike above when converting utf8 to utf16. Is there a reason for this difference? Or should it be changed?